### PR TITLE
fix: support sub day increments for date64

### LINKED ISSUE #6198: Adding sub day seconds to Date64 is ignored.
**Describe the bug**

The second part of `add_month_day_nano` is seemingly ignored, so if you try to add like an hour to a `Date64Type`, you actually get back the same date.

**To Reproduce**

This test will pass on main, but it shouldn't (added to `arrow-array/src/types.rs`).

```rust
#[cfg(test)]
mod tests {
    use super::*;
    use arrow_data::{layout, BufferSpec};

    #[test]
    fn add_month_day_nano() {
        let date_64 = Date64Type::default_value();
        let interval = IntervalMonthDayNanoType::make_value(0, 0, 1_000_000_000 * 3600);

        let result = Date64Type::add_month_day_nano(date_64, interval);

        // result equals date_64, but should be an hour after.
        assert_eq!(result, date_64);
    }
}
```

**Expected behavior**

For date64s adding increments less than a day should be supported.

**Additional context**

I _think_ the result is from coercion to a NaiveDate instead of a NaiveDateTime. I have a working branch, I'll open a PR for... https://github.com/apache/arrow-rs/compare/master...tshauck:add-sub-day-increments-to-date64type?expand=1

### PR DESCRIPTION
**Pull request by tshauck**
# Which issue does this PR close?

Closes #6198

# Rationale for this change
 
It should be possible to add sub day seconds to a date64, but currently those are ignored.

# What changes are included in this PR?

Additional functionality to handle going from a date64type to and from a naivedatetime type. Also a couple of unittests.

# Are there any user-facing changes?

No

### CODE REVIEW

**Comment by tshauck** [comment]
I'm marking this ready for review, but I'm hoping a maintainer can take a quick look to see if this is reasonable, and if so, I'll spend some time updating the tests as they're not testing the seconds component very well after the changes.

File: arrow-array/src/types.rs [hunk 1]
```diff
@@ -1035,6 +1035,16 @@ impl Date64Type {
         epoch.add(Duration::try_milliseconds(i).unwrap())
     }
 
+    /// Converts an arrow Date64Type into a chrono::NaiveDateTime
```
Comment (tshauck) [inline arrow-array/src/types.rs:1038]: This file has the important changes in the PR.

File: arrow-arith/src/numeric.rs [hunk 2]
```diff
@@ -1448,9 +1441,9 @@ mod tests {
         assert_eq!(result.as_ref(), &a);
 
         let b = IntervalDayTimeArray::from(vec![
-            IntervalDayTimeType::make_value(34, 2),
-            IntervalDayTimeType::make_value(3, -3),
-            IntervalDayTimeType::make_value(-12, 4),
+            IntervalDayTimeType::make_value(34, 0),
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:1444]: Why are these changes needed? Would it be possible to update the expected output too? Or is the issue the display looks nasty with millisecond precision too?

I see the same values are used in the tests below as well

File: arrow-arith/src/numeric.rs [hunk 3]
```diff
@@ -1402,16 +1402,13 @@ mod tests {
...
     }
 
-    fn test_date_impl<T: ArrowPrimitiveType, F>(f: F)
-    where
-        F: Fn(NaiveDate) -> T::Native,
-        T::Native: TryInto<i64>,
-    {
-        let a = PrimitiveArray::<T>::new(
+    #[test]
+    fn test_date64_impl() {
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:1406]: maybe this should just be called `test_date64` now as it only tests date 64

File: arrow-arith/src/numeric.rs [hunk 4]
```diff
@@ -1520,4 +1499,94 @@ mod tests {
             "Arithmetic overflow: Overflow happened on: 9223372036854775807 - -1"
         );
     }
+
+    #[test]
+    fn test_date32_impl() {
```
Comment (alamb) [inline arrow-arith/src/numeric.rs:1504]: likewise this might make sense to call `test_date32`

File: arrow-array/src/types.rs [hunk 5]
```diff
@@ -1035,6 +1035,16 @@ impl Date64Type {
...
+    ///
+    /// # Arguments
+    ///
+    /// * `i` - The Date64Type to convert
+    pub fn to_naive_datetime(i: <Date64Type as ArrowPrimitiveType>::Native) -> NaiveDateTime {
+        let datetime = NaiveDateTime::default();
+        datetime.add(Duration::try_milliseconds(i).unwrap())
```
Comment (alamb) [inline arrow-array/src/types.rs:1045]: Can we also add some documentation about when this will panic? Or comments explaining why it won't panic 🤔

**Review by alamb — COMMENTED** [review]
Thanks @tshauck 

I did some research on this. Proposal

Even though Arrow stores Date64 as 

> A 64-bit date type representing the elapsed time since UNIX epoch in milliseconds(64 bits).

https://docs.rs/arrow/latest/arrow/array/types/struct.Date64Type.html

The arrow spec says "where the values are evenly divisible by 86400000"

https://github.com/apache/arrow/blob/51d50d750012d3ca04127f6723c4f1e69ff4f5dc/format/Schema.fbs#L248-L249

Thus I am not sure that setting sub-day precision is a good idea as it would permit (and maybe encourage) constructing Date64 that do not actually follows the Arrow spec 🤔

Maybe we can add some comments to `Date64Type` explaining this nuanace and pointing people to other alternatives (like Timestamp) 

Perhaps we

**Comment by tshauck** [comment]
Thanks for your comments @alamb (and patience 😅 ) ... I think I got anchored to the series generation code in datafusion that uses a date32 for generating series for dates.

I think the bug (https://github.com/apache/datafusion/issues/11823) is related to date handling of less than day increments, i.e. a date plus something smaller than a day doesn't add any additional days as it loses the precision so the loop never exits.

I think maybe the proper path for improving this handling looks something like:

1. Close this PR, to your point about `Date64` being increments of day, vs timestamp being semantically correct
2. (optional) Temp fix to datafusion to throw an error when something smaller than a day is used in `generate_series` with a date.
3. Better fix to coerce the date into a timestamp (this would seem to match postgres and duckdb) and support timestamps general (not yet supported in the `GenSeries` UDF (I think there's a ticket for the latter piece of work).
4. (optional) Comments and/or fallible addition subtraction between `Date*` and intervals when they aren't sufficient. E.g. adding a `Date32` with an interval a `MonthDayNano` when only the nano part is present and less than a day, doesn't align with a day, etc. I can also do a bit of digging against the C++ implementation to see if there's anything similar there.

postgres for reference:

```
postgres=# CREATE TABLE a AS SELECT generate_series(date '2020-01-01', date '2020-02-01', interval '2 day');
SELECT 16
postgres=# select column_name, data_type from information_schema.columns WHERE table_name = 'a';
   column_name   |        data_type         
-----------------+--------------------------
 generate_series | timestamp with time zone
(1 row)
```

**Comment by alamb** [comment]
> Thanks for your comments @alamb (and patience 😅 ) ... I think I got anchored to the series generation code in datafusion that uses a date32 for generating series for dates.

Sorry for the delay in responding. Things have been very busy for me the last few days

> I think maybe the proper path for improving this handling looks something like:
> 
> 1. Close this PR, to your point about `Date64` being increments of day, vs timestamp being semantically correct
> 2. (optional) Temp fix to datafusion to throw an error when something smaller than a day is used in `generate_series` with a date.

Yes I think this sounds like a good idea

> 3. Better fix to coerce the date into a timestamp (this would seem to match postgres and duckdb) and support timestamps general (not yet supported in the `GenSeries` UDF (I think there's a ticket for the latter piece of work).

This (using timestamps in gen sereis) also seems reasonable to me -- I suggest we treat it as a separate ticket though as the infinite loop seems like a bug and this part seems more like a feature / improvement. 


> 4. (optional) Comments and/or fallible addition/subtraction between `Date*` and intervals when they aren't sufficient. E.g. adding a `Date32` with an interval a `MonthDayNano` when only the nano part is present and less than a day, doesn't align with a day, etc. I can also do a bit of digging against the C++ implementation to see if there's anything similar there.

Sounds also reasonable

Thanks again for the help.

**Comment by tustvold** [comment]
See also https://github.com/apache/arrow-rs/issues/5288 and the corresponding discussion on the mailing list - https://lists.apache.org/thread/q036r1q3cw5ysn3zkpvljx3s9ho18419

IMO systems shouldn't ever use Date64 other than for compatibility with systems that require it

**Comment by alamb** [comment]
> See also #5288 and the corresponding discussion on the mailing list - https://lists.apache.org/thread/q036r1q3cw5ysn3zkpvljx3s9ho18419
> 
> IMO systems shouldn't ever use Date64 other than for compatibility with systems that require it

I made https://github.com/apache/arrow-rs/pull/6223 and https://github.com/apache/arrow-rs/pull/6224 to improve the docs
